pkg/lockless_circular_buffer/circular_buffer.go

```diff
@@ -0,0 +1,331 @@
+package lockless_circular_buffer
+
+import (
+	"runtime"
+	"time"
+	"unsafe"
+
+	"go.uber.org/atomic"
+)
+
+type CircularBuffer[T any] struct {
+	buffer []T
+	head   *atomic.Uint32
+	tail   *atomic.Uint32
```

Comment:
these are zero initialized automatically so you can avoid the extra indirection and skip initializing with New(0)